Message priority in IBus

tkirill opened the pull request:
Add support of [priority queues](https://www.rabbitmq.com/priority.html) to `IBus`.  Fix #494.

A failing integration test was added in ba961e7.  The test sends multiple messages with different priorities and expects that messages with higher priority will be received first.  Of course, the test doesn't actually set priorities in this commit but it does it in the final version.

A `priority` parameter was added  to `Publish` method of `IBus` in 90578e1 .

A `MaxPriority` property was added  to `SubscriptionConfiguration` in a63a4c8 .

zidad commented:
Thanks a lot! In general @Pliner @micdenny, Do you think it's a good idea to just have something like MessagePublishOptions as an overload to Publish to prevent a huge amount of overloads with all possible (current & future) permutations of RMQ message settings? There's a lot more options, like TTL, maybe PublishDate (instead of FuturePublish) could be in there as well, to make the interface more simple?

reisenberger commented:
@zidad @Pliner @micdenny @tkirill : I had the same thought as Wiebe. 

 In the subscribe area, there are `Subscribe`/`SubscribeAsync<>()` overloads taking an `Action<ISubscriptionConfiguration>`.  I wondered if a similar pattern could be used to add configuration to Publish.  For example, we add a class `IPublishConfiguration` with a method `IPublishConfiguration WithPriority(int priority);`, with the implementation of this method applying the Priority (or whatever) to the instance of `EasyNetQ.MessagePriorities` that will be used during publishing the message ... something like that.  

This could provide the kind of extensibility Wiebe suggests, and follow existing codebase pattern.

zidad commented:
@reisenberger Something like IBus.Publish<T>(T message, Action<IPublishConfiguration> configurator)? :+1: 

I think MassTransit uses a similar pattern (though I haven't looked at that in a couple of years). I'm pretty sure it goes a little bit against @mikehadlow of the opinionated super-duper happy path simple predefined patterns, but I like the idea of having easy access to all possibilities RMQ offers...

reisenberger commented:
@zidad , yes exactly.  Same pattern already in Subscribe.  To balance simplicity/happy-path against access-to-features,  features opened up can of course be limited by the methods defined on `IPublishConfiguration`.  eg open up `WithPriority()`, maybe `WithExpiration()` etc but not any which would conflict with EasyNetQ's intrinsic way of working (eg delivery-mode, correlationid, reply-to etc - not thought this all through, just examples)

Pliner commented:
@zidad @reisenberger @tkirill Thanks guys!

@tkirill, I absolutely agree with @zidad and @reisenberger, we should escape hell of overloads.

zidad commented:
Not fully implemented yet, but as a start:

https://github.com/zidad/EasyNetQ/blob/add-publish-configuration/Source/EasyNetQ/FluentConfiguration/IPublishConfiguration.cs

Am I missing something, should I add more properties?

micdenny commented:
I agree to have the publish as consume using the configuration fluent interface, **but leave at least the `topic` parameter for backward compatibility**:

```csharp
void Publish<T>(T message, string topic) where T : class;
void Publish<T>(T message, Action<IPublishConfiguration> configurator) where T : class;
```

In this way we're pushing to use the new interface if the developer wants more features, but without breaking anything nor fire warnings. After some releases we could also think to flag the topic one as `[Obsolete("Use fluent configuration interface instead")]`, but I would do it in a second step.

micdenny commented:
> Am I missing something, should I add more properties?

I would start with that, but are you sure we can support multiple topics?

tkirill commented:
Thank you guys for review!  `IPublishConfiguration` is definitely a better option so I replaced `byte?` parameter with `Action<IPublishConfiguration>`.

I implemented only `IPublishConfiguration.WithPriority`.  Overloads with `topic` wasn't removed as @micdenny suggested.

micdenny commented:
I think you can add also Expires and Topic (at least topic), so if someone uses the new fluent api at least have also the option to set the topic, and the expires could be helpful as well (actually I'm doing it using the advanced api so if you add also expires it would reduce my code :smile:)

micdenny commented:
@tkirill I mean you can avoid this `void Publish<T>(T message, string topic, Action<IPublishConfiguration> configure) where T : class;` and just have these overloads:

```csharp
void Publish<T>(T message, string topic) where T : class;
void Publish<T>(T message, Action<IPublishConfiguration> configurator) where T : class;
```

what do you think?

zidad commented:
@micdenny I'm not sure if multiple topics is possible (I didn't try to implement it yet :)), I agree with your suggested overloads (without the topic + action overload).

tkirill commented:
@micdenny Oh, right, you already wrote about removing unnecessary overloads in previous comment and I haven't seen it.  Fixed it in 7ab92f6.

Also, I tried to write docs for `IPublishConfiguration` in a4cec51 but my grammar isn't good so there may be mistakes.

And support for message TTL was implemented in d90d438.  There is a test [Should_delete_message_after_expires_ttl](https://github.com/EasyNetQ/EasyNetQ/commit/d90d43816c97e3d27258a915c8a3d52b913912c1#diff-4f81ef8f31b87cae37c4406d3c344ab7R287).

tkirill commented on Source/EasyNetQ.Tests/FluentConfiguration/PublishConfigurationTests.cs:
```diff
@@ -0,0 +1,36 @@
+        [Test]
+        public void Should_throw_if_default_topic_is_null()
+        {
+            Assert.That(() => new PublishConfiguration(null), Throws.InstanceOf<ArgumentNullException>());
```
Not sure is this test necessary or not.

micdenny replied:
I think yes, because if you don't want to use topic you should not pass
nothing, probably the same as passing null, but in fact WithTopic(null)
does not mean nothing is the default behavior.

Il lunedì 19 ottobre 2015, Kirill Teplinskiy <notifications@github.com> ha
scritto:

> In Source/EasyNetQ.Tests/FluentConfiguration/PublishConfigurationTests.cs
> <https://github.com/EasyNetQ/EasyNetQ/pull/495#discussion_r42360749>:
>
> > @@ -0,0 +1,36 @@
> > +﻿using System;
> > +using EasyNetQ.FluentConfiguration;
> > +using NUnit.Framework;
> > +
> > +namespace EasyNetQ.Tests.FluentConfiguration
> > +{
> > +    [TestFixture]
> > +    public class PublishConfigurationTests
> > +    {
> > +        [Test]
> > +        public void Should_throw_if_default_topic_is_null()
> > +        {
> > +            Assert.That(() => new PublishConfiguration(null), Throws.InstanceOf<ArgumentNullException>());
>
> Not sure is this test necessary or not.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EasyNetQ/EasyNetQ/pull/495/files#r42360749>.
>


-- 
Michael Denny
Engage IT Services s.r.l. <http://www.engageitservices.it/>
Software Architect, Trainer and Consultant
Microsoft MVP .NET <http://mvp.microsoft.com/mvp/Michael%20Denny-5000735>
More about me <http://about.me/micdenny>

tkirill commented on Source/EasyNetQ/RabbitBus.cs:
```diff
@@ -62,25 +69,41 @@ public class RabbitBus : IBus
+                if (configuration.Priority != null)
+                    easyNetQMessage.Properties.Priority = configuration.Priority.Value;
+                if (configuration.Expires != null)
+                    easyNetQMessage.Properties.Expiration = configuration.Expires.ToString();
```
[MessageProperties](https://github.com/EasyNetQ/EasyNetQ/blob/master/Source/EasyNetQ/MessageProperties.cs#L58) has flags `somePropertyPresent`.  These `if`s prevents from setting the flags to `true`.  Is these checks necessary or they could be removed?

micdenny commented:
Thanks, now I'm traveling, late today or tomorrow I'll review it, or
@pliner of course if he can before :)

Il lunedì 19 ottobre 2015, Kirill Teplinskiy <notifications@github.com> ha
scritto:

> @micdenny <https://github.com/micdenny> Oh, right, you already wrote
> about removing unnecessary overloads in previous comment and I haven't seen
> it. Fixed it in 7ab92f6
> <https://github.com/EasyNetQ/EasyNetQ/commit/7ab92f622d1064bf21152f73b07cd251b6c0b88d>
> .
>
> Also, I tried to write docs for IPublishConfiguration in a4cec51
> <https://github.com/EasyNetQ/EasyNetQ/commit/a4cec51fc9a45934acb507a5ad47c33a69ca85dc>
> but my grammar isn't good so there may be mistakes.
>
> And support for message TTL was implemented in d90d438
> <https://github.com/EasyNetQ/EasyNetQ/commit/d90d43816c97e3d27258a915c8a3d52b913912c1>.
> There is a test Should_delete_message_after_expires_ttl
> <https://github.com/EasyNetQ/EasyNetQ/commit/d90d43816c97e3d27258a915c8a3d52b913912c1#diff-4f81ef8f31b87cae37c4406d3c344ab7R287>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EasyNetQ/EasyNetQ/pull/495#issuecomment-149189010>.
>


-- 
Michael Denny
Engage IT Services s.r.l. <http://www.engageitservices.it/>
Software Architect, Trainer and Consultant
Microsoft MVP .NET <http://mvp.microsoft.com/mvp/Michael%20Denny-5000735>
More about me <http://about.me/micdenny>

tkirill commented:
Hello, it is me again!  I see you have a lot of work with EasyNetQ and I just wonder should I fix something in this PR so it could be reviewed when the time comes?  There are merge conflicts at the moment, which way of resolving is preferred in EasyNetQ: rebase or merge?

zidad commented:
@tkirill Thanks! I'm all in favor of rebasing from master to limit the amount of merges/merge commits in the history :+1:

Fydon commented:
Is there a reason why this hasn't been merged yet? Is there a way that I could assist in having this merged?

micdenny commented:
Hi @Fydon I think to now we're waiting a rebase from @tkirill and then a fast review and a merge could be done

zidad commented:
@tkirill If you don't have time I can consolidate/rebase your branch and create a new PR if you prefer?

tkirill commented:
Oops, thanks for reminding me, I've completely forget about this PR.  I'm working on rebasing.  As I understand @zidad correctly I'm supposed to make a new PR after rebase?

zidad commented:
@tkirill Thanks! You don't have to make a new PR after the rebase, only I would have to do that if I'm using your code for a new PR. You will have to _push -f_ though.

tkirill commented:
@zidad I've rebased on the latest master.  Now there is two commits: one for `IPublishConfiguration` and one for `ISubscriptionConfiguration`.  Also, the second commit contains the test.

tkirill commented:
Sorry, I've accidentally reformatted `PublishSubscribeTests.cs`.  I reverted reformatting and pushed again.

tkirill commented on Source/EasyNetQ/RabbitBus.cs:
```diff
@@ -74,20 +91,37 @@ public class RabbitBus : IBus
+        public virtual async Task PublishAsync<T>(T message, Action<IPublishConfiguration> configure) where T : class
+        {
+            Preconditions.CheckNotNull(message, "message");
+            Preconditions.CheckNotNull(configure, "configure");
```
These checks have small technical flaw: they will throw exception only at `await` call.  So this test is failing

    [Test]
        public void Should_throw()
    {
        var testLocalBus = RabbitHutch.CreateBus("host=localhost;prefetchcount=1");
        Assert.Throws<ArgumentNullException>(() => testLocalBus.PublishAsync(default(string), c => { }));
    }

Old `PublishAsync` method had this behavior too and I haven't changed it.  Should it be fixed in a separate PR or it isn't a problem?

micdenny replied:
actually this will throw an exception, but if not awaited will throw an `AggregatedException` that contains an innerExpection = `ArgumentNullException`

we can create async tests, but I don't know if the version of nunit we're using supports it, can you @tkirill open a separated issue for this?

zidad commented:
Hi @tkirill, if you could fix it that would be great! For me it wouldn't have to be a separate PR. Would you mind bumping the version number in version.cs too? 

@micdenny are you OK with merging this PR?

micdenny commented:
checking the PR....

micdenny commented:
ok I like this PR, thank you @tkirill! can you please bump the version file to a breaking version `0.59.0.0`: it is a breaking change because the public api `IBus` has changed.

tkirill commented:
@micdenny Done bump version.

micdenny commented:
@tkirill can you please also add a note for that version, something like:

```
 // 0.59.0.0 Support of priority queues to IBus publish methods
```

tkirill commented:
@micdenny Sure!  Thanks for pointing it out.  I copied your suggested note it seems good :)

micdenny commented:
I'm going to merge, please @tkirill create a separated issue regarding the unit test `Should_throw` about the `PublishAsync` preconditions

zidad commented:
Good stuff gents! Thanks!
